src/control/ros/RosTrajectoryExecutor.cpp

```diff
@@ -0,0 +1,290 @@
+    }
+
+    goal.trajectory.points.emplace_back(waypoint);
+  }
+
+  if (!waitForServer())
+    throw std::runtime_error("Unable to connect to action server.");
+
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
+
+    if (mInProgress)
+      throw std::runtime_error("Another trajectory is in progress.");
+
+    mPromise = std::promise<void>();
+    mInProgress = true;
+    mGoalHandle = mClient.sendGoal(goal, 
+      boost::bind(&RosTrajectoryExecutor::transitionCallback, this, _1));
+
+    return mPromise.get_future();
+  }
+}
+
+//=============================================================================
+bool RosTrajectoryExecutor::waitForServer()
+{
+  using Clock = std::chrono::steady_clock;
+
+  const auto startTime = Clock::now();
+  const auto endTime = startTime + mConnectionTimeout;
+  auto currentTime = startTime + mConnectionPollingRate;
+
+  while (currentTime < endTime)
+  {
+    mCallbackQueue.callAvailable();
+
+    // TODO: Is this thread safe?
+    if (mClient.isServerConnected())
+      return true;
+
+    currentTime += mConnectionPollingRate;
+    std::this_thread::sleep_until(currentTime);
+  }
+
+  return false;
+}
+
+//=============================================================================
+void RosTrajectoryExecutor::transitionCallback(GoalHandle _handle)
+{
+  // This function assumes that mMutex is locked.
+
+  using actionlib::TerminalState;
+  using Result = control_msgs::FollowJointTrajectoryResult;
+
+#if 0
+  ROS_INFO("Transition: CommState = %s TerminalState = %s",
+    _handle.getCommState().toString().c_str(),
+    _handle.getTerminalState().toString().c_str());
+#endif
+  
+  if (_handle.getCommState() == actionlib::CommState::DONE)
+  {
+    std::stringstream message;
+    bool isSuccessful = true;
+
+    // Check the status of the actionlib call. Note that the actionlib call can
+    // succeed, even if execution failed.
+    const auto terminalState = _handle.getTerminalState();
+    if (terminalState != TerminalState::SUCCEEDED)
+    {
+      message << "Action " << terminalState.toString();
+
+      const auto terminalMessage = terminalState.getText();
+      if (!terminalMessage.empty())
+        message << " (" << terminalMessage << ")";
+
+      isSuccessful = false;
+    }
+    else
+    {
+      message << "Execution failed";
+    }
+
+    // Check the status of execution. This is only possible if the actionlib
+    // call succeeded.
+    const auto result = _handle.getResult();
+    if (result && result->error_code != Result::SUCCESSFUL)
+    {
+      message << ": "
+        << getFollowJointTrajectoryErrorMessage(result->error_code);
+
+      if (!result->error_string.empty())
+        message << " (" << result->error_string << ")";
+
+      isSuccessful = false;
+    }
+
+    if (isSuccessful)
+    {
+      mPromise.set_value();
+    }
+    else
+    {
+      mPromise.set_exception(
+        std::make_exception_ptr(
+          std::runtime_error(message.str())));
```

Comment:
We should use a custom `std::runtime_error` subclass here. Otherwise, it will be difficult to distinguish between "expected" (e.g. trajectory was aborted) and "unexpected" failures (e.g. an invalid parameter).